horovod/run/run.py

```diff
@@ -0,0 +1,430 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+import argparse
+import os
+import sys
+import six
+try:
+    from shlex import quote
+except ImportError:
+    from pipes import quote
+
+
+import horovod
+from horovod.run.common.util import codec, safe_shell_exec, timeout, secret
+from horovod.run.driver import driver_service
+from horovod.run.task import task_service
+from horovod.run.util import cache, threads
+
+# Cached information of horovodrun functions be stored in this directory
+CACHE_FOLDER = os.path.join(os.path.expanduser('~'), '.horovod')
+
+# Cache entries will be stale if they are older than this number of minutes
+CACHE_STALENESS_THRESHOLD_MINUTES = 60
+
+# Maximum number of concurrent worker threads to prevent from over saturation.
+MAX_CONCURRENT_EXECUTIONS = 20
+
+# Number of retries for sshing into the hosts
+SSH_RETRIES = 5
+
+
+def use_cache():
+    """
+    If decorates a function, and if cache_disabled is set, it will store the
+        output of the function if it not None. If a function output is None, the
+    execution will not be cached.
+    :return:
+    """
+
+    def wrap(func):
+        def wrap_f(*args, **kwargs):
+            fn_cache = kwargs.pop('fn_cache')
+            if fn_cache is None:
+                results = func(*args, **kwargs)
+            else:
+                cached_result = fn_cache.get(
+                    (func.__name__, tuple(args[0]), frozenset(kwargs.items())))
+                if cached_result is not None:
+                    return cached_result
+                else:
+                    results = func(*args, **kwargs)
+                    if results is not None:
+                        fn_cache.put(
+                            (func.__name__, tuple(args[0]),
+                             frozenset(kwargs.items())),
+                            results)
+            return results
+
+        return wrap_f
+
+    return wrap
+
+
+@use_cache()
+def _check_all_hosts_ssh_successful(host_addresses, ssh_port=None):
+    """
+    checks if ssh can successfully be performed to all the hosts.
+    :param host_addresses: list of addresses to ssh into. for example,
+        ['worker-0','worker-1']
+        ['10.11.11.11', '10.11.11.12']
+    :type host_addresses: list(strings)
+    :return: Returns True if all ssh was successful into all the addresses.
+    """
+
+    def exec_command(command):
+        exit_code = 1
+        msg_stdout = ""
+        msg_stderr = ""
+
+        # Try ssh 5 times
+        for i in range(SSH_RETRIES):
+            stdout_w = six.StringIO()
+            stderr_w = six.StringIO()
```

Comment:
Idea: create a single StringIOs `output`, pass it to both stdout and stderr of execute(), and modify output to look like this:
```
root@d89f89ff64aa:/examples# horovodrun -np 2 -H <ip> python -c 'print("lala")'
ssh not successful for host <ip>:
Permission denied, please try again.
Permission denied, please try again.
Permission denied (publickey,password).

root@d89f89ff64aa:/examples# 
```